Sortable upload files

chrisbuttery opened the pull request:
The main purpose of this PR was to render all uploading and uploaded files into a `Sortable` component.

This [link](https://github.com/icelab/formalist-standard-react/pull/17#issuecomment-214635721) references a list of feedback that needed addressing - with at least 50% of it already merged in to master.

![sortable](https://cloud.githubusercontent.com/assets/121675/15277661/39460980-1b4e-11e6-9fdc-6131687ee1a6.gif)

Whenever you get a chance @makenosound, I'll be happy to go through it with you. 👍

makenosound commented:
Hey @chrisbuttery 

Ta for the PR! Have pulled it down and run through a few things. I think there are a few polish bits we can do, and some errors/one additional feature I think we want to resolve:

## UI/UX feedback

- [ ] The drop target UI should obscure the existing files when you’re dragging. It’s too complex (visually) and you don’t need to know about the existing files while you’re dragging. As soon as you drag over the page it should look _identical_ to the state then there are no files.
- [ ] The dotted borders are a bit fat I think, they could prolly be a pixel smaller?
- [ ] The dotted borders are truncated on the right for me.
- [ ] If you have existing uploads, and you drag over the "Drop image here" label (or the border), the drop-ness triggers on/off so the UI starts flashing like crazy.
- [ ] The files "fade in" on first load (with existing data) which feels a little awkward to me.
- [ ] Newly added files to a multiple are added at the top, but then move to the bottom when they complete. This feels awkward, I think I may have suggested it be that way, but I think they should ideally just get represented in the order they’re added (so new ones go to the bottom).
- [ ] I think the fade transitions are all a little slow, would be good to speed them up a touch.

## Errors

- [ ] I got an error trying to remove a file while it was uploading. Seems to consistently fail for me.

## Features

- [ ] Have we addressed the way we’d pass in a custom component yet? It’s something we’ll want pretty soon, so may as well wrap it up with this push. I can talk you through how it works in other spots (well, the one other spot).

Cheers,
Max

chrisbuttery commented:
![components](https://cloud.githubusercontent.com/assets/121675/15345028/eecb73f2-1cee-11e6-8de3-bcd0f7b8c030.gif)

I've refactored the labels and border for the dropzones to prevent the glitchy effect.
Existing files are hidden on drag of the window.
I've removed the fade-in animation. This was introduced as a hacky way of avoiding the late rendering of a thumbnail by the browser - but the length of the animation was too long to 'feel right'.
The error task should now be redundant. You'll just have to make sure you've wiped your node_modules first.

@makenosound Happy to go through this with you, when you get some time.

makenosound commented:
I have looked and have another list for ye:

- [ ] The drop zones should *not* change size when dragging new files over. They should hide the files as though they are _behind_ the "Drop image here" overlay.
- [ ] I get some `cannot read property 'forEach' of undefined` errors for each file I have, looks like it’s in `renderCustomTemplate`.
- [ ] Each uploaded file seems to have a `cursor: pointer` on it, but you can’t actually do anything with them.
- [ ] You got the order backwards for new files added to a `multiUploadField`, I said "I think they should ideally just get represented in the order they’re added (so new ones go to the bottom)." but it’s the opposite. The reason they should go to the bottom is so the general order is preserved, since it _could_ be meaningful.
- [ ] The alignment of the 'Upload an image' button to the field label is a bit awkward, I think we need to bring the button down so the text is closer to matching baselines.
- [ ] I totally trashed all my `node_modules` in both `formalist-demo` and `formalist-standard-react` and I’m still getting that error when trying to remove a file _while_ it’s uploading.
- [ ] The initial, empty state for the dropzone is very, very light — it’s almost too hard to read on my home monitor. I think we need to bump it up a fair bit.
- [ ] It looks like _sometimes_ when dragging you still get a funny mixed state where the "Drop image" stuff and the existing uploaded files show together.

chrisbuttery commented:
@makenosound Thanks for the feedback.
I'm pretty sure I've addressed everything above.

I now add an overlay to file-upload fields that will slightly obscure existing files.

I could not re-create the error with removing files.
I've tried referencing npm hosted and local versions of `attache-upload` and it's performing as expected.

![blah](https://cloud.githubusercontent.com/assets/121675/15457398/e2da1cfc-20cb-11e6-83eb-6a2db5fa5205.gif)

makenosound commented:
@chrisbuttery — nice work! That’s feeling better for sure. Couple of things that I missed last time:

- [ ] The little height jump when switching from uploading to uploaded feels a little crappy. Reckon it’s possible to get them to be the same height? Would mean adjusting the thumbnail a little I think, but that’s probably better than having the entire form below shift around.
- [ ] The `buttonText` prop isn’t actually settable from the AST. It’s something we’d need to pass through on the `attributes` object, and I think we want to have it’s name be more in line with the way we’ve done it in the `selectionField`. There it’s `selector_label`, so we want something similar, perhaps `upload_label`.
- [ ] We also want the dropzone hint to be configurable in the same way (I noticed it’s hardcoded in CSS now?), perhaps `dropzone_label` for that.

Those last two will require an adjustment on the ruby side of things to allow them to come through, but I can do that.

chrisbuttery commented:
@makenosound Shouldn't be an issue 👍 

The`dropzone_label` _was_ configurable yeah, before I moved it into a pseudo element. All good

chrisbuttery commented:
![foo](https://cloud.githubusercontent.com/assets/121675/15463518/558fb5cc-210d-11e6-8e98-a569542053d7.gif)

Now rendering custom button and label text

chrisbuttery commented:
![height](https://cloud.githubusercontent.com/assets/121675/15463564/a79c7710-210d-11e6-99f3-05ab8fd4a8f7.gif)

Consistent height for uploading/uploaded files

chrisbuttery commented:
@makenosound Those latest feedback items have been addressed.

I'll merge those other formalist-* `bumped-node-deps` PRs once this one is finalised 👍

chrisbuttery commented:
Damn. I've noticed the label is now showing when items are then sorted. _sigh_. 
Must've broke something when switching from a pseudo element -> `<span>`.
One sec.

makenosound commented:
👏  Nice work — all those things I’d noted before look all good. Have seen a couple of regressions or things I didn’t notice before though:

- [ ] The sortable content disappears for me when sorting the list (which doesn’t happen in the `multiSelectionField` for example)
- [ ] I get a + cursor (like when copying a file in Finder) when trying to sort the list.

Bugs:

- [ ] I dropped a file onto a multi-field with existing data (loaded from the server) and those files were lost. Looks like it’s happening for all new uploads (regardless of how they’re triggered).

chrisbuttery commented:
Hmm OK cool. I'll get on it.

chrisbuttery commented:
@makenosound The issue with elements disappearing from an dragged Sortable Item has been addressed.
The problem was that if the wrapping parent of a `Sortable` with `position: relative` then the `<Items/>` would "lose" elements on drag. Very strange.
The parent was positioned relative so that the dropzone label and button could then be positioned absolute.
The solution involved refactoring how the dropzone indicates it's a dropzone (such as the background colour and dashed border ).

The bug with existing files being clobbered on upload was also addressed.
`multiple` is defined by default props (to allow multiple files), however a recent changed the script to oncly checks `this.props.attributes` for a `multiple` value.
So now the script will check both, as a single file upload with passing `multiple={true}` through props.

The green copy icon on drag is tricky.
It seems to occasionally flicker on the `multiSelectionField` too, yet sticks around on this Dropzone. It must be an event listener in the Dropzone 3rd party component - but I'm not sure as yet.

chrisbuttery commented:
This last change overrides the `effectAllowed` from `"all"` to `"move"` on DragStart.
When set to `"all"` all operations are permitted - meaning the element can be moved, copied etc, hence the change in cursor.

![dragstart](https://cloud.githubusercontent.com/assets/121675/15637186/7ee4cb44-2656-11e6-8ec1-d5ef75dc7fdf.gif)

makenosound commented:
Cool, those things seem to be resolved. I’m seeing a bit of weirdness around the drag-and-drop behaviour though:

![](http://files.icelab.com.au/drops/max/1464574161-weirdness.gif)

See how the entire field disappears sometimes? This seems to be reproducible in nested and non-nested fields by switching between Finder and the browser window while dragging.

I’m also seeing these errors in the console:

```
Uncaught Invariant Violation: Expected targetIds to be an array

Uncaught Invariant Violation: Cannot call hover while not dragging.

Warning: setState(...): Can only update a mounted or mounting component. This usually means you called setState() on an unmounted component. This is a no-op. Please check the code for the DropZone component.
```

chrisbuttery commented:
@makenosound Yeah I noticed that weird disappearing act this morning.
It should've been address with 5f63173.

What was happening is that if you drop immediately on the field from Finder it triggers and active state on the dropzone as well as a hover state - which I had a rule declared to hide the dropzone label.

This rule should've specified `if the dropzone 'has children' and was active and then hovered over - hide the label`. Which was added in c3db702

chrisbuttery commented:
@makenosound Just an update to let you know that I'm looking into those 2 `Uncaught Invariant Violation` errors. They are emitted from `react-dnd-html5-backend` which is invoked from `Sortable`.

chrisbuttery commented:
Created a PR to solve this conflict.
https://github.com/gaearon/react-dnd-html5-backend/issues/32

chrisbuttery commented:
We may have a weiner.

![hmm](https://cloud.githubusercontent.com/assets/121675/15736373/b720ac2e-28e2-11e6-8228-adfdbce1cf41.gif)

makenosound commented on src/components/fields/multi-upload-field/index.js:
```diff
@@ -823,30 +863,24 @@ const MultiUploadField = React.createClass({
+              ? this.renderFiles(files)
               : null}
           </Dropzone>
         </div>
```
@chrisbuttery Just a note that we need to include the `<FieldErrors/>` component here as per all the other field types.

chrisbuttery commented:
@makenosound `<FieldErrors />` module included 👍

makenosound commented:
@chrisbuttery Those weirdnesses seems to be gone! Have a few more things I came across in testing this time. Mostly pretty minor:

- [ ] In the provided data, the `file_name` property has been changed to `name`. Was that something we did on purpose? I vaguely recall some discussion about making it `file_name` so it was really obvious. I’m not fussed either way, but perhaps @timriley has thoughts? We’d need to upload ALS if we do change it is all.
- [ ] If I drag two files on to upload at the same time and the first one completes before the second they swap order. And once the second one completed I get a funny cached render where both have the same thumbnail for a few seconds.
- [ ] If you drag a new file onto the singular `upload_field` while there is a file uploading you get _two_ files uploading/uploaded.
- [ ] The cursor disappears entirely when hovering the "Drop file to upload" label.
- [ ] The cursor doesn’t change to a hand when hovering the dropzone.

timriley commented:
> In the provided data, the file_name property has been changed to name. Was that something we did on purpose? I vaguely recall some discussion about making it file_name so it was really obvious. I’m not fussed either way, but perhaps @timriley has thoughts? We’d need to upload ALS if we do change it is all.

I'd be in favour of keeping it at `file_name`. It probably got moved about as I focused discussion on the general behaviour, rather than the specifics of the required data keys (and maintaining backwards compat)

chrisbuttery commented:
When a file has officially been updated this script will now:
1. clone and update the fileObject that has been uploaded
2. clone the existing files and find the index of a file that has a matching `uid`.
3. remove the file at that index (which is a preview file) and replace it with our cloned 'uploaded' file
4. save to state

Originally i was filtering the array of files and then pushing the new fileObject onto the end - which was screwing with the order.

![persist_index](https://cloud.githubusercontent.com/assets/121675/15848399/28b4878e-2cd0-11e6-91d7-e3dea4b2ccdd.gif)

makenosound commented:
@chrisbuttery Looks good to me! 💯
